rama-net/src/client/pool/mod.rs

```diff
@@ -716,58 +746,77 @@ where
+                };
+
+                let pool_result = if let Some(duration) = self.wait_for_pool_timeout {
+                    timeout(duration, pool.get_conn(&conn_id))
                     .await
                     .map_err(|err|{
                         trace!("pooled connector: timeout triggered while waiting for a connection (/w conn id: {conn_id:?}) from pool");
                         OpaqueError::from_std(err)
                     })?
-            } else {
-                pool.get_conn(&conn_id).await
+                } else {
+                    pool.get_conn(&conn_id).await
+                };
+
+                match pool_result? {
+                    ConnectionResult::Connection(conn) => {
+                        trace!(
+                            "pooled connector: got connection (w/ conn id: {conn_id:?}) from pool (running healtch check now)"
+                        );
+                        // TODO timeout on this or global timeout vs only wait_for_pool_timeout?
+                        match conn.health_check().await {
+                            Ok(conn) => {
+                                trace!("pooled connector: connection is detected as health");
+                                return Ok(EstablishedClientConnection { conn, input });
+                            }
+                            Err(err) => {
+                                trace!(?err, "pooled connector: connection was not health");
+                                continue;
+                            }
+                        };
+                    }
+                    ConnectionResult::CreatePermit(permit) => {
+                        trace!(
+                            "pooled connector: no connection (w/ conn id: {conn_id:?}) found, received permit to create a new one"
+                        );
+                        permit
+                    }
+                }
             };
 
-            match pool_result? {
-                ConnectionResult::Connection(c) => {
-                    trace!(
-                        "pooled connector: got connection (w/ conn id: {conn_id:?}) from pool, returning"
-                    );
-                    return Ok(EstablishedClientConnection { conn: c, input });
-                }
-                ConnectionResult::CreatePermit(permit) => {
-                    trace!(
-                        "pooled connector: no connection (w/ conn id: {conn_id:?}) found, received permit to create a new one"
-                    );
-                    permit
-                }
-            }
-        };
+            let EstablishedClientConnection { input, conn } =
+                self.inner.connect(input).await.map_err(Into::into)?;
 
-        let EstablishedClientConnection { input, conn } =
-            self.inner.connect(input).await.map_err(Into::into)?;
+            // TODO here we create a new connection, but do we also want to health check this for consistency,
```

Comment:
I think it's fair to assume that the inner connection only returns a solid connector back? Because otherwise even if it passes _now_, what about 1 sec later 🤷